tracer/src/Datadog.Trace/Span.cs

```diff
@@ -545,5 +548,12 @@ private void WriteCloseDebugMessage()
                 "Span closed: [s_id: {SpanId}, p_id: {ParentId}, t_id: {TraceId}] for (Service: {ServiceName}, Resource: {ResourceName}, Operation: {OperationName}, Tags: [{Tags}])\nDetails:{ToString}",
                 new object[] { Context.RawSpanId, Context.ParentIdInternal, Context.RawTraceId, ServiceName, ResourceName, OperationName, Tags, ToString() });
         }
+
+        internal void AddSpanLink(Span spanLinkToAdd, List<KeyValuePair<string, object>> optionalAttributes = null)
```

Comment:
Thinking a little about how someone would use this API, maybe we should remove the `optionalAttributes` parameter and return the `SpanLink` instead of `void`. Then the caller can add attributes to the `SpanLink`. Just brainstorming here. The calling code might look like:
```csharp
var spanLink = span.AddSpanLink(otherSpan);
spanLink.AddAttribute(name1, value1);
spanLink.AddAttribute(name2, value2);
```

Reply:
oh this would also make it chainable right?
```var spanLink = span.AddSpanLink(otherSpan).AddAttribute(name1, value1);```

can make AddAttribute just add to the KV list we have - should be a simple change.

Reply:
> this would also make it chainable right?

You could go even further and have `SpanLink.AddAttribute()` also return the `SpanLink` (`return this;`).
```csharp
var spanLink = span.AddSpanLink(otherSpan)
                   .AddAttribute(name1, value1)
                   .AddAttribute(name2, value2);
```

We did this with `Span.AddTag()` a long time ago, but I'm not sure if we actually like that today 😅 

@andrewlock wdyt?

Reply:
it's up king

Reply:
ok it's up for real now.

Reply:
gentle reping 👍🏾

Reply:
Copying from my other comment:

I think the `return null` is a tad risky in a fluent-like interface that `Span.AddSpanLink` and `SpanLink.AddAttribute` have.

I'd see the method signature and think that I should be able to do the following with a fluent interface like this ->

`span.AddSpanLink(someOtherSpan, someAttributes).AddAttribute(someKey, someValue).AddAttribute(someKey2, someValue2);`

but if `span` is marked as `IsFinished` I'll get a `null` everywhere so I have to guard everything with `?.` ->

`span.AddSpanLink(someOtherSpan, someAttributes)?.AddAttribute(someKey, someValue)?.AddAttribute(someKey2, someValue2);`


I'm wondering if changing the return type back to `void` would be the best now for both `AddSpanLink` and `AddAttribute` now that they have that `IsFinished` call.

Reply:
leaving `AddSpanLink` the same but `AddAttribute` to return void.